[ONNX] Support complex in FX exporter

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106140
* __->__ #100554

Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

titaiwangms committed:
[ONNX] Support complex in FX exporter

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -216,6 +216,17 @@ def _fill_tensor_shape_type(
+        elif expected_value.dtype in _type_utils.COMPLEX_TO_FLOAT:
+            # Like torch.view_as_real, we flatten complex tensors to real tensors with
+            # additional last dimension of 2
+            onnxscript_value.shape = tuple(
+                [dim if isinstance(dim, int) else None for dim in expected_value.size()]
+                + [2]
+            )
```
```suggestion
            onnxscript_value.shape = (
                *[dim if isinstance(dim, int) else None for dim in expected_value.size()],
                2
            )
```

justinchuby commented on torch/onnx/_type_utils.py:
```diff
@@ -359,6 +359,11 @@ def valid_torch_name(torch_name: Union[TorchName, str]) -> bool:
     float: {"tensor(float16)", "tensor(float)", "tensor(double)"},
     bool: {"tensor(int32)", "tensor(int64)", "tensor(bool)"},
 }
+COMPLEX_TO_FLOAT = {
```
I would add a comment for this mapping

justinchuby commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -194,6 +194,9 @@ def generate_random_tensors(shape: torch.Size, dtype: torch.dtype):
         )
+    if dtype in _type_utils.COMPLEX_TO_FLOAT:
+        # ONNX does not support complex value, but support real representation
+        return torch.randn(shape + (2,), dtype=_type_utils.COMPLEX_TO_FLOAT[dtype])
```
```suggestion
        return torch.randn((*shape, 2), dtype=_type_utils.COMPLEX_TO_FLOAT[dtype])
```

justinchuby reviewed (dismissed):
Awesome!

justinchuby commented on torch/onnx/_type_utils.py:
```diff
@@ -359,6 +359,11 @@ def valid_torch_name(torch_name: Union[TorchName, str]) -> bool:
     float: {"tensor(float16)", "tensor(float)", "tensor(double)"},
     bool: {"tensor(int32)", "tensor(int64)", "tensor(bool)"},
 }
+COMPLEX_TO_FLOAT = {
```
Good call

titaiwangms replied:
https://github.com/pytorch/pytorch/pull/103942

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

thiagocrepaldi reviewed (dismissed):
lgtm. the remaining comments are not blocker, just suggestions

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -136,11 +137,18 @@
         # "nn.functional.scaled_dot_product_attention"  non-deterministic
         "scatter_add",
         "scatter_reduce",
+        "stft",
```
Should `stft` be put in `COMPLEX_TESTED_OPS`?

titaiwangms replied:
Sorry, this PR needs registry updated. I will turn it to draft for now.

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"




[ghstack-poisoned]

justinchuby commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -618,6 +630,51 @@ def _should_skip_xfail_test_sample(
+    dtype: torch.dtype,
+    op: opinfo_core.OpInfo,
+):
+    # device is provided by instantiate_device_type_tests, but we only want to run in cpu.
```
Right, I don't think we need to limit ourselves to cpus here, unlike in torchlib. However, I think it's also ok to assume transitivity in `torch gpu` == `torch cpu` == `onnx cpu`? Unless we want to test ORT cuda here.

justinchuby commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -595,6 +608,51 @@ def _should_skip_xfail_test_sample(
+                model = SingleOpModel(op.op, cpu_sample.kwargs)
+                model.eval()
+
+                if dtype == torch.float32:
+                    # Relax atol and rtol for float32 based on empirical results
+                    rtol = 1e-5
+                    atol = 2e-5
+                else:
+                    rtol = None
+                    atol = None
+                # Run the test
```
```suggestion
```

Use the tolerance field in OpInfo instead?

titaiwangms replied:
Found it's strict. So I added tasks to https://github.com/pytorch/pytorch/issues/105338. But it's doable.

justinchuby commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -627,46 +685,15 @@ class TestOnnxModelOutputConsistency(onnx_test_common._TestONNXRuntime):
+        allowed_dtypes=onnx_test_common.COMPLEX_TYPES,
+    )
+    def test_output_match_complex(self, device: str, dtype: torch.dtype, op):
+        """Test the ONNX exporter."""
```
I would expand this docstring a little.

justinchuby commented on torch/onnx/_internal/fx/registration.py:
```diff
@@ -26,12 +26,14 @@ class SymbolicFunction:
     onnx_function: The symbolic function from torchlib.
     is_custom: Whether the function is a custom function.
+    is_complex: Whether the function is a complex function.
 
```
```suggestion
```

nit

justinchuby commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -238,6 +238,11 @@ def generate_random_tensors(shape: torch.Size, dtype: torch.dtype):
             random_numbers > 0.5, torch.tensor(True), torch.tensor(False)
         )
+    if fx_type_utils.is_torch_complex_dtype(dtype):
+        # ONNX does not support complex value, but support real representation
```
```suggestion
        # ONNX does not support complex values but supports their real representations
```

nit: grammar for readability

justinchuby commented on torch/onnx/_internal/fx/registration.py:
```diff
@@ -112,6 +125,7 @@ def register_custom_op(
             op_name: The name of the operator to register.
             overload: The overload of the operator to register. If it's default overload,
                 leave it to None.
+            is_complex: Whether the function is a complex function.
```
```suggestion
            is_complex: Whether the function is a function that handles complex inputs.
```

titaiwangms commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -133,6 +138,64 @@ def dispatch(
 
+    @_beartype.beartype
+    @diagnostics.diagnose_call(
+        diagnostics.rules.find_operator_overloads_in_onnx_registry,
```
@BowenBao Does it make sense to have two functions calling the same rule?

BowenBao replied:
Not really. Each op will populate two diagnostics of this same rule, but one for this function and another for `get_function_overloads`.

titaiwangms replied:
Moved it down to `get_function_overloads`

justinchuby commented on torch/onnx/_internal/fx/type_utils.py:
```diff
@@ -30,9 +30,11 @@ def dtype(self) -> Optional[torch.dtype]:
 
 
-def is_torch_complex_dtype(tensor: TensorLike) -> bool:
+def is_torch_complex_dtype(tensor_or_dtype: TensorLike | torch.dtype) -> bool:
```
```suggestion
def is_torch_complex_dtype(tensor_or_dtype: torch.dtype) -> bool:
```

justinchuby commented on torch/onnx/_internal/fx/type_utils.py:
```diff
@@ -30,9 +30,11 @@ def dtype(self) -> Optional[torch.dtype]:
+def is_torch_complex_dtype(tensor_or_dtype: TensorLike | torch.dtype) -> bool:
     # NOTE: This is needed as TorchScriptTensor is nor supported by torch.is_complex()
-    return tensor.dtype in _COMPLEX_TO_FLOAT
+    if isinstance(tensor_or_dtype, torch.dtype):
+        return tensor_or_dtype in _COMPLEX_TO_FLOAT
+    return tensor_or_dtype.dtype in _COMPLEX_TO_FLOAT
```
```suggestion
    return tensor_or_dtype in _COMPLEX_TO_FLOAT
```

to reduce confusion? i.e. Users don't need to consider whether to use `is_torch_complex_dtype(val)` or `is_torch_complex_dtype.dtype`. Ideally there should only be one correct way.

justinchuby commented:
LGTM. Just a few minor comments

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"


Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"


Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

[ghstack-poisoned]

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -395,11 +402,17 @@ def _compare_pytorch_onnx_with_ort(
+        # NOTE: ONNX Runtime doesn't support tensor of complex64/complex128, so we
+        # convert them to float32/float64.
+        # TODO: Need stft enabled to test complex64/complex128
+        if isinstance(ref_output, torch.Tensor) and torch.is_complex(ref_output):
```
You should add a new input/output adaptor step for this. Checkout `RemoveNonTensorInputStep` for example.
Feel free to do it in this or follow-up PR.

titaiwangms replied:
Done

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -344,9 +344,16 @@ def run_ort(
+        else input
+        for input in pytorch_inputs
+    ]
+    ort_input = {k: v.cpu().numpy() for k, v in zip(input_names, pytorch_inputs)}
```
ditto: input adapter step

titaiwangms replied:
Done

BowenBao reviewed (approved):
LGTM w/ comments. Please also resolve others' comments before landing.

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"


Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

[ghstack-poisoned]

justinchuby commented:
Are we handling any complex constants that may be inside the graph? Maybe that’s an unlikely case that we don’t need to care yet?

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"


Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

[ghstack-poisoned]

titaiwangms commented:
> Are we handling any complex constants that may be inside the graph? Maybe that’s an unlikely case that we don’t need to care yet?

It's not covered in this PR. We can add it when we have a model or use case.

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"


Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support complex in FX exporter"


Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge

titaiwangms commented:
@pytorchbot merge
